devlib/target.py

```diff
@@ -279,6 +280,11 @@ def shutils(self):
             self._setup_shutils()
         return self._shutils
 
+    def is_running(self, process):
+        cmd = f'''{self.busybox} ps -A -o comm,stat | {self.busybox} awk '$1 == "{process}" && $2 != "Z"' '''
```

Comment:
This breaks for a task named with a space, such as:
```
$ ../devlib/devlib/bin/x86_64/busybox ps -A -o comm,stat | grep foo 
foo bar.sh       S  
```
Because `$2` will be set to "bar.sh".
Instead, you need to use `stat,comm` so that $1 is always the state (which AFAIK can never contain a space), and then match on the rest by emptying $1 (note that `$0` has a leading space in the value, probably a leftover separator from $1/$2, so we need to add it to `{process}` before matching):
```
        cmd = f'''{self.busybox} ps -A -o stat,comm | {self.busybox} awk 'alive=($1 != "Z"); $1=""; $alive && $0 == " {process}"'''
```

Reply:
Ah right, bit of an academic distinction as there's exactly 0 tasks either on my Android phone or laptop that would have a space in the comm name so this is pretty much exclusive to the "I'll make a script to create this edge case" scenario but sure still worth fixing.
The command
```
/data/local/tmp/bin/busybox ps -A -o stat,comm | /data/local/tmp/bin/busybox awk 'alive=($1 != "Z"); $1=""; $alive && $0 == " traced"'
```
however just returns all the tasks in the system, are you sure there's not an issue here somewhere?

Reply:
> Ah right, bit of an academic distinction

It's unfortunately not. I never hit the issue, _except_ on Android devices: https://bugzilla.kernel.org/show_bug.cgi?id=210193
The problematic comm was `shell srvc 1574`

A quick study on some WA example traces from 2021 I have on my laptop show:
```
AsyncTask #1
AsyncTask #3
DataSender Main
DataSender Send
GLThread 107
GLThread 113
Jit thread pool
magnifier pixel
netlink socket
Profile Saver
Runtime worker 
shell svc 6791
shell svc 6800
Shutdown thread
```

A trace from april 2023 from a pixel 6 shows:
```
AsyncTask #1
shell svc 14342
shell svc 14346
shell svc 14350
shell svc 14354
shell svc 14381
shell svc 14385
```

>  so this is pretty much exclusive to the "I'll make a script to create this edge case" scenario but sure still worth fixing.

After a few years you will realize that a good deal of the alleged "edge cases" are just a convenient statement to sweep issues under the rug. 1/2 of my contributions to devlib have been to fix real-world problems coming out of this sort of shortcuts. 1/4 of my contributions have been about fixing real-world problems coming out of similar code, but wrote as an oversight (including by myself) rather than a conscious decision to ignore the problem.

When you pile up all the features devlib has and how much of that is used in a real-world use case like WA, you quickly realize that if you don't design things to be robust on purpose, you will have problems literally all the time. That used to be the situation with LISA, and to some extent devlib, and a lot of work has been done to align everything and ensure robustness, so these days I place the bar on robustness of the code pretty high for new stuff, as we can't really afford coming back to everything a month later when we realize it's broken.


> however just returns all the tasks in the system, are you sure there's not an issue here somewhere?

Yes, I had a typo that made it not match anything. I can't come up with something that actually work, so I'd suggest just getting the output and processing in Python, as it will be trivial. Something along the line:
```
process = process.strip()
def parse(row):
    state, comm = row.split(' ', 1)
    return state.strip(), comm.strip()
[
    comm.
    for state, comm in map(parse, out.splitlines())
    if comm == process and state != "Z" 
]
```

Reply:
Ah actually here we go, this seems to work:
```
busybox ps -A -o stat,comm | busybox awk 'BEGIN{found=0} {state=$1; $1=""; if ($state != "Z" && $0 == " zsh") {found=1}} END {print found}'
```
f-string can still contain literal `{` and `}` if you use `{{`  `}}`

Reply:
I'm very happy for us to get this corner case fixed of course and you're right on the merit but the things you're listing are output from `trace-cmd`, not output from `busybox ps`. Things like `AsyncTask #1`  that my GB6 traces are full of are threads of a process called `labs.geekbench6` and the only thing that shows up in the `ps` output is the latter.

The help option for `busybox ps` gives this:
```
oriole:/data/local/tmp/bin # ./busybox ps --help
BusyBox v1.32.1 (2021-04-15 16:06:54 BST) multi-call binary.

Usage: ps [-o COL1,COL2=HEADER] [-T]

Show list of processes

	-o COL1,COL2=HEADER	Select columns for display
	-T			Show threads

```
Since we're not passing the -T option, these threads will not show up in the output. Even when passing the -T option while running an iteration of GB6 I still could not find even one field that'd have a space in the `comm` but that's a different story. 

Regardless, the approach you suggested works so I updated the PR.

Reply:
> Since we're not passing the -T option

Maybe we should ? I did not really think about threads vs process as the distinction is a bit of a POSIX compat technicality in Linux (not completely but mostly). I'd rather ere on the side of listing more than less, otherwise a user might want to match on a thread name, and not realize the test will always return false.

Reply:
Sure we might as well, now that we have the spaces in the names handled as well it can't hurt. Should be good to go now.

Reply:
LGTM